商品購入機能実装

ryotechcamp opened the pull request:
# What
・API導入
・クレジットカード決済での購入機能を実装する
・モデルの単体テストコードを書く。

# Why
商品購入機能を実装するため

# Gyazo
・必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://i.gyazo.com/58d242eecd6222762c579858cfd0621c.mp4

・入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://i.gyazo.com/b958cd2fe0fa0850db5765e30ed46912.mp4

・ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://i.gyazo.com/49643dcc5f5bf2adb0b141bac3e71fe6.mp4

・ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://i.gyazo.com/09dc7d845beffe0d4562def061d2e9eb.mp4

・ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://i.gyazo.com/f28af259959f29282cbd50bcc9c4b984.mp4

・売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://i.gyazo.com/c874eda9a17ae60ae3379d7c18b5ae48.mp4

・売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://i.gyazo.com/62f13ad0261a450092ec930cbab4fc0b.mp4

・ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://i.gyazo.com/12e01ad25cca34d4c5f7c0395bc3d755.mp4

・ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
(ryoというユーザーがryoの編集ページへ遷移しようとしています)
https://i.gyazo.com/bd14cd9839db19f0400f3764b3f32765.mp4

・テスト結果の画像
https://i.gyazo.com/4e711a5361349f94070ea41926d8d495.png

ryotechcamp committed:
フォームオブジェクト作成

ryotechcamp committed:
テーブル作成

購入情報(purchase)テーブルと配送先情報(shipping_address)テーブルもモデル及びテーブルを作成した

ryotechcamp committed:
テーブル間のアソシエーション定義

ryotechcamp committed:
フォームオブジェクト保存機能追記

ryotechcamp committed:
ルーティング・アクション設定

ryotechcamp committed:
indexアクション中身記載

ryotechcamp committed:
indexファイル配置

ryotechcamp committed:
リンク先パス指定

ryotechcamp committed:
フォームオブジェクト修正

ryotechcamp committed:
indexアクション追加

ryotechcamp committed:
テーブル名モデル名変更

ryotechcamp committed:
エラー文表示

ryotechcamp committed:
テストコード記述

ryotechcamp committed:
ファイル名変更

ryotechcamp committed:
エラー表示文記述

ryotechcamp committed:
カード情報ビューファイル変更

ryotechcamp committed:
turbolink無効化

ryotechcamp committed:
payjp読み込み

ryotechcamp committed:
ファイル名変更

ryotechcamp committed:
card.js作成

ryotechcamp committed:
テストコードtoken追加

ryotechcamp committed:
pay.jp追加

ryotechcamp committed:
webpacker追加

ryotechcamp committed:
環境変数導入

ryotechcamp committed:
条件分岐

yutamasumoto commented on spec/models/purchase_address_spec.rb:
```diff
@@ -0,0 +1,77 @@
+        @purchase_address.valid?
+        expect(@purchase_address.errors.full_messages).to include("Telephone number can't be blank")
+      end
+      it '電話番号が全角では保存できないこと' do
```
こちら電話番号に関しましては`電話番号は11桁以内の数値のみ保存可能なこと（09012345678となる）`の実装項目があるかとございます。
こちらを満たすために下記のテストコードを実装して頂ければと思います。
```
電話番号は半角英字混合だと登録できない
```

yutamasumoto commented on spec/models/purchase_address_spec.rb:
```diff
@@ -0,0 +1,77 @@
+      end
+    end
+
+    context '内容に問題がある場合' do
```
こちら現状ユーザー情報の有無、商品情報の有無に関するテストが行われておりませんので追記してみましょう。

yutamasumoto commented on spec/models/purchase_address_spec.rb:
```diff
@@ -0,0 +1,77 @@
+      it '郵便番号が空だと保存できないこと' do
+        @purchase_address.postal_code = ''
+        @purchase_address.valid?
+        #binding.pry
```
```
#binding.pry
```
こちらのコメントアウトですが、不要であれば削除しましょう。
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめします。

yutamasumoto reviewed (commented):
```
ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
(ryoというユーザーがryoの編集ページへ遷移しようとしています)
https://i.gyazo.com/bd14cd9839db19f0400f3764b3f32765.mp4
```
上記挙動動画についてのコメントとなります。
`ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移すること`の実装要件に関しまして`displays_controller.rb`に適切な条件分岐の記載が必要となるのですが確認させていただいたプルリクエストの差分内には`displays_controller.rb`の変更履歴を確認することができませんでした。
こちらcommit・push忘れでしたら再度commit・pushして頂き、`displays_controller.rb`にて条件分岐を追記されていませんでしたら`ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移すること`の実装要件に即した条件分岐を再度`displays_controller.rb`にて追記して頂ければと存じます。

ryotechcamp committed:
条件分岐

ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移すること
（pushし忘れていました。）

ryotechcamp committed:
電話番号テストコード

「電話番号は半角英字混合だと登録できない」を実装しました

ryotechcamp committed:
ユーザー情報の有無、商品情報の有無に関するテスト

ryotechcamp committed:
コメントアウト消去

ryotechcamp commented:
修正いたしました。
「ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移すること」の実装要件はpushし忘れです。失礼いたしました。

saki125 commented on app/views/displays/show.html.erb:
```diff
@@ -24,12 +26,12 @@
 
     <% if user_signed_in? %>
-      <% if current_user.id == @display.user_id %>
+      <% if current_user.id == @display.user_id && @display.purchase.blank? %>
```
```ruby
<% if current_user.id == @display.user_id && @display.purchase.blank? %>
```
こちらの条件ですと、前の条件が不一致であればelse以下が読まれます。そのため、出品者じゃない場合売り切れていても購入のボタンが出るのではないでしょうか？
以下のように記述していただくことで防げるかと存じます。
```ruby
if "ログインしていれば" かつ "商品は売れていなければ" 
  if "現在ログインしているユーザーが商品の出品者であれば"
    商品の編集
    削除
  else
    購入画面に進む
  end
end
```

saki125 reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加で１点コメントいたしました。非常に細かい部分で恐縮ではございますが、ご確認いただけますと幸いです。
修正後再度、レビュー依頼フォームのご提出をお願いいたします。
お手数をお掛け致しますが、よろしくお願いいたします。

ryotechcamp committed:
条件分岐修正

ryotechcamp commented:
ご指摘いただきありがとうございます。
修正いたしました。

saki125 commented:
レビュー依頼ありがとうございます！
LGTMとさせていただきます。実装お疲れさまでした。

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切るようにしましょう。
引き続き実装頑張ってください！応援しております。
